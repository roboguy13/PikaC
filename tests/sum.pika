%generate sum

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

sum : Sll -> Int;
sum Nil := 0;
sum (Cons h t) := h + (sum t);

%test "sum" Int: sum (Cons 1 (Cons 2 (Nil [Sll]) [Sll]) [Sll]);

