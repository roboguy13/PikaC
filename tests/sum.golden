void sum(loc x, loc i1);
void sum(loc x, loc i1) {
 if (x == 0) {
  WRITE_LOC(i1, 0, 0);
 } else {
  loc h = READ_LOC(x, 0);
  loc n = READ_LOC(x, 1);
  sum(n, i1);
  loc i = READ_LOC(i1, 0);
  WRITE_LOC(i1, 0, (long)h + (long)i);
 }
}